### PR TITLE
Add Graph Node ID to Raw Node ID remapping support

### PR DESCRIPTION
**Pull request by classicsong**
#508 

*Description of changes:*

By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

### CODE REVIEW

File: python/graphstorm/gconstruct/remap_result.py [hunk 1]
```diff
@@ -0,0 +1,809 @@
...
+    logging.info("{%d} Remapping edge predictions takes {%f} secs", rank, dur)
+
+
+def remap_edge_pred(pred_etypes, pred_dir,
+                    output_dir, out_chunk_size,
+                    num_proc, rank, world_size, with_shared_fs,
+                    preserve_input=False):
+    """ Remap edge prediction result.
+
+        The function wil iterate all the edge types that
```
Comment (zhjwy9343) [inline python/graphstorm/gconstruct/remap_result.py:388]: wil -> will

File: python/graphstorm/model/utils.py [hunk 2]
```diff
@@ -601,6 +697,172 @@ def save_hdf5_embeddings(emb_path, embeddings, rank, world_size,
...
+        #     emb.part00001.bin
+        #     ...
+
+        The content of nids.part files and emb.part files looks like:
+
+        Example:
+        --------
+
+        .. code::
+        nids.part00000.bin   |   emb.part00000.bin
```
Comment (zhjwy9343) [inline python/graphstorm/model/utils.py:808]: 1. add a space line after `.. code::`;
2. add indent to each line of the contents.

File: python/graphstorm/model/utils.py [hunk 3]
```diff
@@ -601,6 +697,172 @@ def save_hdf5_embeddings(emb_path, embeddings, rank, world_size,
...
+        23                   |   0.472,0.432,-0.732,...
+        ...
+
+        Note: `save_pytorch_embeddings` (called by `save_embeddings`) is different from
+        `save_full_node_embeddings`. In `save_pytorch_embeddings`, it will shuffle the
+        order of node embeddings so that the node embeddings are shuffled according to
+        node IDs in Graph Node ID space. While `save_full_node_embeddings`
+        shuffles node IDs instead of node embeddings, which is more efficient.
+
+        Note: User need to call graphstorm.gcostruct.remap_result to remap the output
```
Comment (zhjwy9343) [inline python/graphstorm/model/utils.py:822]: User -> Users

File: python/graphstorm/model/utils.py [hunk 4]
```diff
@@ -601,6 +697,172 @@ def save_hdf5_embeddings(emb_path, embeddings, rank, world_size,
...
+        "Only support save embeddings in the format of ['pytorch']"
+    ntypes = list(embeddings.keys())
+    nid_shuffler = NodeIDShuffler(g, node_id_mapping_file, ntypes) \
+                if node_id_mapping_file else None
+
+    pb = g.get_partition_book()
+    shuffled_embs = {}
+    for ntype in ntypes:
+        if get_rank() == 0:
+            logging.info("save embeddings pf %s to %s", ntype, save_embed_path)
```
Comment (zhjwy9343) [inline python/graphstorm/model/utils.py:851]: pf -> of?

**Comment by zhjwy9343** [comment]
A general issue is about the suffix of saved artifacts. Sometime tensors are saved as *.bin, and sometime as *.pt. Are there any reason to use the two suffix names?

File: python/graphstorm/model/utils.py [hunk 5]
```diff
@@ -705,6 +1101,132 @@ def save_prediction_results(predictions, prediction_path, rank):
...
 
+def save_node_prediction_results(predictions, prediction_path):
+    """ Save node predictions to the given path
+
+        The saved node prediction results looks like:
+
+        Example:
+        --------
+        .. code::
+            PATH_TO_RESULTS:
```
Comment (zhjwy9343) [inline python/graphstorm/model/utils.py:1112]: Add a space line after `.. code::`.

File: python/graphstorm/model/utils.py [hunk 6]
```diff
@@ -705,6 +1101,132 @@ def save_prediction_results(predictions, prediction_path, rank):
...
+
+        The result_info.json contains three information:
+           * "format", how data are stored, e.g., "pytorch".
+           * "world_size", the total number of file parts. 0 means there is no partition.
+           * "ntypes", a list of node types that have prediction results.
+
+        Example:
+        --------
+        .. code::
+            {
```
Comment (zhjwy9343) [inline python/graphstorm/model/utils.py:1132]: ditto

File: python/graphstorm/model/utils.py [hunk 7]
```diff
@@ -705,6 +1101,132 @@ def save_prediction_results(predictions, prediction_path, rank):
...
+                "format": "pytorch",
+                "world_size": 8,
+                "ntypes": ["movie", "user"]
+            }
+
+        .. note::
+        The saved prediction results are in Graph Node ID space.
```
Comment (zhjwy9343) [inline python/graphstorm/model/utils.py:1139]: Add indents in each line of the note contents.

File: python/graphstorm/model/utils.py [hunk 8]
```diff
@@ -705,6 +1101,132 @@ def save_prediction_results(predictions, prediction_path, rank):
...
+
+def save_edge_prediction_results(predictions, prediction_path):
+    """ Save edge predictions to the given path
+
+        Example:
+        --------
+        The saved node prediction results looks like:
+
+        .. code::
+            PATH_TO_RESULTS:
```
Comment (zhjwy9343) [inline python/graphstorm/model/utils.py:1173]: Add a space line after `.. code::`.

File: python/graphstorm/model/utils.py [hunk 9]
```diff
@@ -705,6 +1101,132 @@ def save_prediction_results(predictions, prediction_path, rank):
...
+
+        The result_info.json contains three information:
+           * "format", how data are stored, e.g., "pytorch".
+           * "world_size", the total number of file parts. 0 means there is no partition.
+           * "etypes", a list of edge types that have prediction results.
+
+        Example:
+        --------
+        .. code::
+            {
```
Comment (zhjwy9343) [inline python/graphstorm/model/utils.py:1196]: ditto

File: python/graphstorm/model/utils.py [hunk 10]
```diff
@@ -705,6 +1101,132 @@ def save_prediction_results(predictions, prediction_path, rank):
...
+                "format": "pytorch",
+                "world_size": 8,
+                "etypes": [("movie","rated-by","user"), ("user","watched","movie")]
+            }
+
+        .. note::
+        The saved prediction results are in Graph Node ID space.
```
Comment (zhjwy9343) [inline python/graphstorm/model/utils.py:1203]: Add indent in the each line of the note contents.

File: python/graphstorm/gconstruct/remap_result.py [hunk 11]
```diff
@@ -0,0 +1,809 @@
...
+        f"World size must be larger than 0, but get {world_size}."
+    assert rank < world_size, \
+        f"Expecting {world_size} workers but the worker ID is {rank}."
+    out_chunk_size = args.output_chunk_size
+    assert out_chunk_size > 0, \
+        f"Output chunk size should be larger than 0 but get {out_chunk_size}."
+
+    ################## remap embeddings #############
+    emb_ntypes = []
+    if node_emb_dir is not None:
```
[the remark this answers is not in the document]
Comment (classicsong) [reply python/graphstorm/gconstruct/remap_result.py:591]: No. It is possible that there is no node embeddings that need to be remapped.

File: python/graphstorm/gconstruct/remap_result.py [hunk 12]
```diff
@@ -0,0 +1,809 @@
...
+        # If node embedding exists, we are going to remap all the embeddings.
+        if with_shared_fs:
+            if os.path.exists(os.path.join(node_emb_dir, "emb_info.json")):
```
[the remark this answers is not in the document]
Comment (classicsong) [reply python/graphstorm/gconstruct/remap_result.py:594]: When with_shared_fs, emb_info.json should be there. If it does not exists, we just skip remapping the node embeddings.

File: python/graphstorm/model/utils.py [hunk 13]
```diff
@@ -456,21 +459,56 @@ def remap_embeddings(embeddings, rank, world_size,
...
+        # mode 767 means rwx-rw-rwx:
+        #     - owner of the folder can read, write, and execute;
+        #     - owner' group can read, write;
+        #     - others can read, write, and execute.
+        os.chmod(emb_path, 0o767)
+
+    # make sure the emb_path permission is changed before other process start to save
+    barrier()
+
+    assert rank < world_size
```
[the remark this answers is not in the document]
Comment (classicsong) [reply python/graphstorm/model/utils.py:503]: Usually, not.

**Comment by classicsong** [comment]
> A general issue is about the suffix of saved artifacts. Sometime tensors are saved as *.bin, and sometime as *.pt. Are there any reason to use the two suffix names?

Updated

**Review by zhjwy9343 — APPROVED** [review]
Please push to update the last two comments. Then, 

LGTM.

**Review by zheng-da — APPROVED** [review]

**Pull request merged**